File: c/3/maxprimef.c
#include "maxprimef.h"

unsigned int max_prime_fac(long long num) {
    unsigned int max_prime = 0;
    unsigned int div = 2;
    while (num > 1) {
        while (num % div == 0) {            
            if (div > max_prime) max_prime = div;
            num /=  div;
        }
        div += 1;
    }
    return max_prime;
}

File: haskell/3/maxprimef.hs
prime_factors :: (Integral a) => a -> [a]
prime_factors n =
    case factors of
        [] -> [n]
        _  -> factors ++ prime_factors (n `div` (head factors))
    where factors = take 1 $ filter (\x -> (n `mod` x) == 0) [2..(n `div` 2)]

max_prime_fac :: Integer -> Integer
max_prime_fac n = maximum (prime_factors n)
